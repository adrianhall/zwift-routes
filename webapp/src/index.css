body, div#root {
  height: 100%;
  left: 0;
  margin: 0;
  padding: 0;
  position: fixed;
  top: 0;
  width: 100%;
}
