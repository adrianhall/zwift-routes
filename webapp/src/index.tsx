import React from 'react';
import ReactDOM from 'react-dom';

const Application: React.SFC<{}> = () => (
  <h1>In Application</h1>
);

ReactDOM.render(<Application />, document.getElementById('root'));
