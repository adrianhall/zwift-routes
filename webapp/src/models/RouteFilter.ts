import ZwiftSport from './ZwiftSport';

/**
 * Definition of the filters that can be applied to the 
 * route list.
 */
export default interface RouteFilter {
  /**
   * Filter by the Zwift World.  If set, matches zwiftWorld.
   */
  world?: string;

  /**
   * Filter by the Sport.  If set, matches 
   */
  sport?: ZwiftSport;

  /**
   * Include completed routes in the set.
   */
  includeCompletedRoutes: boolean;

  /**
   * Include event only routes in the set.
   */
  includeEventOnlyRoutes: boolean;
  
  /**
   * Include the default world (Watopia) in the list.
   */
  includeDefaultWorld: boolean;

  /**
   * Include if the route if above the minimum route.
   */
  maximumZwiftLevel?: number;
}
