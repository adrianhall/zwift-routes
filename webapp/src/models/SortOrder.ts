import SortField from "./SortField";

interface SortOrder {
  /**
   * What field within the Zwift Route to order by
   */
  orderBy: SortField;

  /**
   * Ascending (true) or Descening?
   */
  ascending: boolean;
}

export default SortOrder;
