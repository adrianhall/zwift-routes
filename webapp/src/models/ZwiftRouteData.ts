/**
 * Definition for the static user data that is stored in routes.json
 */
export default interface ZwiftRouteData {
  /**
   * The unique ID for the route.
   */
  id: string;

  /**
   * The VeloViewer difficulty rating
   */
  difficulty: number;

  /**
   * The route distance (in km).
   */
  distance: number;

  /**
   * The elevation gain for the route (in m).
   */
  elevation: number;

  /**
   * True if this route is available only in events.
   */
  eventonly: boolean;

  /**
   * The distance from the drop-point to the route start (in km).
   */
  leadin_distance: number;

  /**
   * The elevation gain from the drop-point to the route start (in m).
   */
  leadin_elevation: number;

  /**
   * Minimum Zwift level required to use this route (default: 0)
   */
  level?: number;

  /**
   * Optional link to the Zwift Insider page for this route.
   */
  link?: string;
  
  /**
   * Name of the route on Zwift.
   */
  name: string;

  /**
   * The sports that this route applies to.
   */
  sports?: string;

  /**
   * The Zwift World that holds this route.
   */
  world: string;
}


