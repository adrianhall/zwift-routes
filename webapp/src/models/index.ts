import capitalize from 'lodash/capitalize';

/**
 * Possible values for the display-units.  Either imperial (miles and feet),
 * or metric (kilometers and meters).
 */
export enum DisplayUnits {
  Imperial            = 'imperial',
  Metric              = 'metric'
}

/**
 * Possible values for the list sort direction - ascending or descending.
 */
export enum SortDirection {
  Ascending           = 'ascending',
  Descending          = 'descending'
}

/**
 * Possible values for the sort field.
 */
export enum SortField { 
  Name                = 'name',
  Difficulty          = 'difficulty',
  RouteDistance       = 'routeDistance',
  RouteElevationGain  = 'routeElevationGain',
  TotalDistance       = 'totalDistance',
  TotalElevationGain  = 'totalElevationGain'
}

/**
 * List of potential sports that can be done on Zwift.
 */
export enum ZwiftSport {
  AllSports           = '*',
  Cycling             = 'cycling',
  Running             = 'running'
}

/**
 * Definition of the display settings to use.
 */
export interface DisplaySettings {
  units: DisplayUnits;
}

/**
 * Definition of a filter used to limit the number of routes in the list.
 */
export interface RouteFilter {
  world?: string;
  sport?: ZwiftSport;
  includeCompletedRoutes: boolean;
  includeEventOnlyRoutes: boolean;
  includeDefaultWorld: boolean;
  maximumZwiftLevel?: number;
}

/**
 * When we sort, we sort by a field and direction.
 */
export interface SortOrder {
  sortField: SortField;
  sortDirection: SortDirection;
}

/**
 * The format of a single routes.json entry.
 */
export interface ZwiftRouteData {
  id: string;
  difficulty: number;
  distance: number;
  elevation: number;
  eventonly: boolean;
  leadin_distance: number;
  leadin_elevation: number;
  level?: number;
  link?: string;
  name: string;
  sports?: string;
  world: string;
}

/**
 * The format of the data written to / read from the IndexedDB
 */
export interface ZwiftUserData {
  routeId: string;
  isCompleted: boolean;
}

/**
 * A ZwiftRoute combines the static data from the routes.json file and
 * the dynamic data from the IndexedDB.
 */
export class ZwiftRoute {
  readonly id: string;
  
  readonly difficulty: number;
  isCompleted: boolean;
  readonly isEventOnly: boolean;
  readonly isForCycling: boolean;
  readonly isForRunning: boolean;
  readonly leadinDistance: number;
  readonly leadinElevationGain: number;
  readonly minimumZwiftLevel: number;
  readonly routeDistance: number;
  readonly routeElevationGain: number;
  readonly routeName: string;
  readonly totalDistance: number;
  readonly totalElevationGain: number;
  readonly zwiftInsiderLink?: URL;
  readonly zwiftWorld: string;

  constructor(staticData: ZwiftRouteData, dynamicData?: ZwiftUserData) {
    const sport = (staticData.sports || 'all').toLowerCase();

    this.id = staticData.id;
    this.difficulty = staticData.difficulty;
    this.isCompleted = dynamicData?.isCompleted || false;
    this.isEventOnly = staticData.eventonly;
    this.isForCycling = sport === 'all' || sport === 'cycling';
    this.isForRunning = sport === 'all' || sport === 'running';
    this.leadinDistance = staticData.leadin_distance;
    this.leadinElevationGain = staticData.leadin_elevation;
    this.minimumZwiftLevel = staticData.level || 0;
    this.routeDistance = staticData.distance;
    this.routeElevationGain = staticData.elevation;
    this.routeName = staticData.name;
    this.totalDistance = staticData.leadin_distance + staticData.distance;
    this.totalElevationGain = staticData.leadin_elevation + staticData.elevation;

    const w = staticData.world.split(/\s+/);
    this.zwiftWorld = w.map((word) => capitalize(word)).join(' ');

    if (staticData.link) {
      this.zwiftInsiderLink = new URL(staticData.link);
    }
  }

  /**
   * Gets the dynamic data for a route.
   */
  getZwiftUserData(): ZwiftUserData {
    return { routeId: this.id, isCompleted: this.isCompleted };
  }

  /**
   * Checks to see if a route needs to be included or excluded based on 
   * filter criteria.
   * 
   * @param filter the RouteFilter to use
   * @returns true if it is to be included; false if excluded.
   */
  isMatch(filter: RouteFilter): boolean {
    if (filter.world && filter.world !== '*') {
      const worldSet = [ filter.world ];
      if (filter.includeDefaultWorld) worldSet.push('Watopia');
      if (!worldSet.includes(this.zwiftWorld)) return false;
    }

    if (filter.sport && filter.sport !== ZwiftSport.AllSports) {
      if (filter.sport === ZwiftSport.Cycling && !this.isForCycling) return false;
      if (filter.sport === ZwiftSport.Running && !this.isForRunning) return false;
    }

    if (filter.maximumZwiftLevel) {
      if (this.minimumZwiftLevel >= filter.maximumZwiftLevel) return false;
    }

    if (!filter.includeCompletedRoutes && this.isCompleted) return false;
    if (!filter.includeEventOnlyRoutes && this.isEventOnly) return false;

    return true;
  }
}

