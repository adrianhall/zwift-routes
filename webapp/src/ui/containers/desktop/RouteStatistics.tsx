import React from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Typography from '@material-ui/core/Typography';
import { ZwiftRoute } from 'src/models';
import uniq from 'lodash/uniq';

const useStyles = makeStyles((theme: Theme) => createStyles({
  root: {
    padding: theme.spacing(2),
    width: '94%'
  },
  table: {
    marginTop: theme.spacing(2)
  }
}));

interface RouteStats {
  completedCount: number;
  totalCount: number;
  completedScore: number;
  totalScore: number;
}

const emptyStats = {
  completedCount: 0,
  totalCount: 0,
  completedScore: 0,
  totalScore: 0
};

const updateStats = (obj: RouteStats, route: ZwiftRoute): void => {
  obj.totalCount += 1;
  obj.totalScore += route.difficulty;
  if (route.isCompleted) {
    obj.completedCount += 1;
    obj.completedScore += route.difficulty;
  }
}

export interface RouteStatisticsProps {
  routes: ZwiftRoute[];
}

const RouteStatistics: React.SFC<RouteStatisticsProps> = 
  ({ routes }: RouteStatisticsProps) => {
    const styles = useStyles();
    const worlds = uniq(routes.map((r) => r.zwiftWorld)).sort();

    const routeStats: Record<string,RouteStats> = {};
    worlds.forEach((world) => {
      routeStats[world] = { ...emptyStats };
    });
    const totals: RouteStats = { ...emptyStats };

    routes.forEach((route) => {
      updateStats(totals, route);
      updateStats(routeStats[route.zwiftWorld], route);
    });

    const completedStats = (stats: RouteStats): string => {
      const percent = (stats.completedCount / stats.totalCount) * 100.0;
      return `${percent.toFixed(1)}%`;
    };

    const scoreStats = (stats: RouteStats): string => {
      return stats.completedScore.toFixed(2);
    };
    
    return (
      <div className={styles.root}>
        <Typography align="center" variant="h5">Route Statistics</Typography>
        <TableContainer component={Paper} className={styles.table}>
          <Table>
            <TableHead>
              <TableRow>
                <TableCell>Route</TableCell>
                <TableCell align="center">Completed</TableCell>
                <TableCell align="center">Score</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {worlds.map((world) => (
                <TableRow key={world}>
                  <TableCell>{world}</TableCell>
                  <TableCell align="center">{completedStats(routeStats[world])}</TableCell>
                  <TableCell align="center">{scoreStats(routeStats[world])}</TableCell>
                </TableRow>
              ))}
              <TableRow>
                <TableCell><b>Totals</b></TableCell>
                <TableCell align="center"><b>{completedStats(totals)}</b></TableCell>
                <TableCell align="center"><b>{scoreStats(totals)}</b></TableCell>
              </TableRow>
            </TableBody>
          </Table>
        </TableContainer>
      </div>
    );
  };

export default RouteStatistics;
