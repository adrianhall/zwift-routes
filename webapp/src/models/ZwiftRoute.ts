import capitalize from 'lodash/capitalize';
import ZwiftUserData from './ZwiftUserData';
import RouteFilter from './RouteFilter';
import ZwiftSport from './ZwiftSport';
import ZwiftRouteData from './ZwiftRouteData'

/**
 * The model for a single ZwiftRoute
 */
export default class ZwiftRoute {
  private staticData: ZwiftRouteData;
  private dynamicData: ZwiftUserData;
  private sport: string;

  constructor(staticData: ZwiftRouteData, dynamicData?: ZwiftUserData) {
    this.staticData = staticData;
    this.dynamicData = dynamicData || { routeId: this.staticData.id, isCompleted: false };
    this.sport = this.staticData.sports || 'all';
  }
  
  /**
   * Capitalize each word in a string.
   * 
   * @param s the input string
   */
  private capWords(s: string): string {
    const w = s.split(/\s+/);
    return w.map((word) => capitalize(word)).join(' ');
  }

  /**
   * Returns the unique ID for the route
   */
  get id(): string {
    return this.staticData.id;
  }

  /**
   * Returns the VeloViewer difficulty rating.
   */
  get difficulty(): number {
    return this.staticData.difficulty;
  }

  /**
   * Returns true if the user has completed the ride
   */
  get isCompleted(): boolean {
    return this.dynamicData.isCompleted;
  }

  /**
   * Sets or clears the user completion flag
   */
  set isCompleted(value: boolean) {
    this.dynamicData.isCompleted = value;
  }

  /**
   * Returns true if this route can only be used during events.
   */
  get isEventOnly(): boolean {
    return this.staticData.eventonly;
  }

  /**
   * Returns true if you can ride a bike on this route.
   */
  get isForCycling(): boolean {
    return this.sport === 'all' || this.sport === 'cycling';
  }

  /**
   * Returns true if you can run on this route.
   */
  get isForRunning(): boolean {
    return this.sport === 'all' || this.sport === 'running';
  }

  /**
   * Returns the distance from the drop-point to the route start (in km).
   */
  get leadinDistance(): number {
    return this.staticData.leadin_distance;
  }

  /**
   * Returns the elevation gain from the drop-point to the route start (in m).
   */
  get leadinElevationGain(): number {
    return this.staticData.leadin_elevation;
  }

  /**
   * The minimum level required to use this route.
   */
  get minimumZwiftLevel(): number {
    return this.staticData.level || 0;
  }

  /**
   * The official distance (start to end) of the route (in km).
   */
  get routeDistance(): number {
    return this.staticData.distance;
  }

  /**
   * The official elevation gain (start to end) of the route (in m).
   */
  get routeElevationGain(): number {
    return this.staticData.elevation;
  }

  /**
   * The official name of the route.
   */
  get routeName(): string {
    return this.staticData.name;
  }

  /**
   * The total distance (including lead-in) needed to complete this ride, in km.
   */
  get totalDistance(): number {
    return this.staticData.leadin_distance + this.staticData.distance;
  }

  /**
   * The total elevation gain (including lead-in) needed to complete 
   * this ride, in m.
   */
  get totalElevationGain(): number {
    return this.staticData.leadin_elevation + this.staticData.elevation;
  }

  /**
   * The URL to the Zwift Insider route description, or undefined if it
   * doesn't exist.
   */
  get zwiftInsiderLink(): string | undefined {
    return this.staticData.link;
  }

  /**
   * The name of the world.  We normalize the world name such that each
   * world name is capitalized, to allow for proper sorting and comparison.
   */
  get zwiftWorld(): string {
    return this.capWords(this.staticData.world);
  }

  /**
   * Returns the ZwiftUserData for this object.  This is used when storing
   * the data.
   */
  getZwiftUserData(): ZwiftUserData {
    return {
      routeId: this.id,
      isCompleted: this.isCompleted
    };
  }

  /**
   * Determines if this route is a match for the provided filter
   * 
   * @param filter the route filter to compare this route to
   * @returns true if this route matches the filter.
   */
  isMatch(filter: RouteFilter): boolean {
    // The general process is to escape out (return false) if something
    // doesn't match - if everything matches, then return true.
    if (filter.world) {
      const worldSet = [ filter.world ];
      if (filter.includeDefaultWorld) worldSet.push('Watopia');
      if (!worldSet.includes(this.zwiftWorld)) {
        return false;
      }
    }

    if (filter.sport && filter.sport === ZwiftSport.Cycling && !this.isForCycling) {
      return false;
    }

    if (filter.sport && filter.sport === ZwiftSport.Running && !this.isForRunning) {
      return false;
    }

    if (!filter.includeCompletedRoutes && this.isCompleted) {
      return false;
    }

    if (!filter.includeEventOnlyRoutes && this.isEventOnly) {
      return false;
    }

    if (filter.maximumZwiftLevel && this.minimumZwiftLevel >= filter.maximumZwiftLevel) {
      return false;
    }
    
    return true;
  }
}
