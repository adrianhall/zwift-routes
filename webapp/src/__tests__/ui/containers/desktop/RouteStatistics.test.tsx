/* eslint-env jest */
import React from 'react';
import { render } from '@testing-library/react';
import RouteStatistics from 'src/ui/containers/desktop/RouteStatistics';
import routeData from 'src/data/routes.json';
import { ZwiftRoute } from 'src/models';

const routes: ZwiftRoute[] = routeData.map((r, i) => {
  const route = new ZwiftRoute(r);
  if (i % 2 === 0) route.isCompleted = true;
  return route;
});

describe('desktop', () => {
  describe('RouteStatistics', () => {
    // Snapshot test
    it('should render like the snapshot', () => {
      const { container } = render(<RouteStatistics routes={routes} />);
      expect(container).toMatchSnapshot();
    });
  });
});
