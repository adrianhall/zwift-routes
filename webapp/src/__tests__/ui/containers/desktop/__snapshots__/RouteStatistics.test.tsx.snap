// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`desktop RouteStatistics should render like the snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
    >
      Route Statistics
    </h5>
    <div
      class="MuiPaper-root MuiTableContainer-root makeStyles-table-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        class="MuiTable-root"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Route
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
              scope="col"
            >
              Completed
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
              scope="col"
            >
              Score
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Bologna
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              100.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              5.88
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Crit City
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              0.49
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Innsbruck
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              52.94
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              London
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              148.91
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              New York
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              53.60
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Richmond
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              42.9%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              14.23
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Watopia
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.9%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              371.34
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Yorkshire
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              50.0%
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              12.21
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <b>
                Totals
              </b>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              <b>
                50.4%
              </b>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            >
              <b>
                659.60
              </b>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
